File: surf-cloud-core/src/main/java/dev/slne/surf/cloud/SurfCloudMainApplication.java
package dev.slne.surf.cloud;

import dev.slne.surf.cloud.api.SurfCloudApplication;
import dev.slne.surf.cloud.api.exceptions.FatalSurfError;
import java.lang.reflect.Method;
import java.util.concurrent.TimeUnit;
import lombok.extern.flogger.Flogger;
import org.apache.commons.lang3.ArrayUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
import org.springframework.beans.BeansException;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.core.NestedRuntimeException;
import org.springframework.scheduling.annotation.AsyncConfigurer;

@Flogger
@SurfCloudApplication(
    jpaBasePackages = "dev.slne.surf.cloud.core.test",
    redisBasePackages = {"dev.slne.surf.cloud.core.util"}
)
public class SurfCloudMainApplication implements AsyncConfigurer, ApplicationContextAware {

  private ApplicationContext applicationContext;

  @Override
  public AsyncUncaughtExceptionHandler getAsyncUncaughtExceptionHandler() {
    return (ex, method, params) -> {

      if (ex instanceof FatalSurfError fatalSurfError) {
        handleFatalSurfError(fatalSurfError, method, params);
      } else if (ex instanceof NestedRuntimeException nested
          && nested.getRootCause() instanceof FatalSurfError fatalSurfError) {
        handleFatalSurfError(fatalSurfError, method, params);
      } else {
        log.atSevere()
            .atMostEvery(1, TimeUnit.SECONDS)
            .withCause(ex)
            .log(
                """
                    Exception message - %s
                    Method name - %s
                    ParameterValues - %s
                    """,
                ex.getMessage(),
                method.getName(),
                ArrayUtils.toString(params)
            );
      }
    };
  }


  @Override
  public void setApplicationContext(@NotNull ApplicationContext applicationContext) throws BeansException {
    this.applicationContext = applicationContext;
  }

  private void handleFatalSurfError(@NotNull FatalSurfError fatalSurfError, @NotNull Method method, Object[] params) {
    log.atSevere()
        .log(
            "Fatal error occurred in method %s with parameters %s",
            method.getName(),
            ArrayUtils.toString(params)
        );

    log.atSevere()
        .log(fatalSurfError.buildMessage());

    if (applicationContext != null) {
      SpringApplication.exit(applicationContext, fatalSurfError);
    }
  }
}

File: surf-cloud-core/surf-cloud-core-common/src/main/kotlin/dev/slne/surf/cloud/SurfCloudMainApplication.kt
package dev.slne.surf.cloud

import dev.slne.surf.cloud.api.common.SurfCloudApplication
import dev.slne.surf.cloud.api.common.exceptions.FatalSurfError
import dev.slne.surf.surfapi.core.api.util.logger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asExecutor
import org.apache.commons.lang3.ArrayUtils
import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.AutoConfigurationPackage
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.core.NestedRuntimeException
import org.springframework.scheduling.annotation.AsyncConfigurer
import java.lang.reflect.Method

@SurfCloudApplication
@AutoConfigurationPackage
class SurfCloudMainApplication : AsyncConfigurer, ApplicationContextAware {
    private val log = logger()
    private lateinit var applicationContext: ApplicationContext

    override fun setApplicationContext(applicationContext: ApplicationContext) {
        this.applicationContext = applicationContext
    }

    override fun getAsyncExecutor() = Dispatchers.IO.asExecutor()

    override fun getAsyncUncaughtExceptionHandler() =
        AsyncUncaughtExceptionHandler { ex, method, params ->
            if (ex is FatalSurfError) {
                handleFatalSurfError(ex, method, params)
            } else if (ex is NestedRuntimeException && ex.rootCause is FatalSurfError) {
                handleFatalSurfError(ex.rootCause as FatalSurfError, method, params)
            } else {
                log.atSevere()
                    .withCause(ex)
                    .log(
                        """
                            Exception message - %s
                            Method name - %s
                            ParameterValues - %s
                            """.trimIndent(),
                        ex.message,
                        method.name,
                        ArrayUtils.toString(params)
                    )
            }
        }


    private fun handleFatalSurfError(
        fatalSurfError: FatalSurfError,
        method: Method,
        params: Array<Any>
    ) {
        log.atSevere()
            .log(
                "Fatal error occurred in method %s with parameters %s",
                method.name,
                ArrayUtils.toString(params)
            )

        log.atSevere()
            .log(fatalSurfError.buildMessage())

        SpringApplication.exit(applicationContext, fatalSurfError)
    }
}